controllers/dockermachine_controller.go

```diff
@@ -47,17 +68,257 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 
 	dockerMachine := &infrastructurev1alpha2.DockerMachine{}
 	if err := r.Client.Get(ctx, req.NamespacedName, dockerMachine); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
 		log.Error(err, "failed to get dockerMachine")
+		return ctrl.Result{}, err
+	}
+
+	// Find the owner reference
+	var machineRef *metav1.OwnerReference
+	for _, ref := range dockerMachine.OwnerReferences {
+		if ref.Kind == machineKind.Kind && ref.APIVersion == machineKind.Version {
+			machineRef = &ref
+			break
+		}
+	}
+	if machineRef == nil {
+		log.Info("did not find matching machine reference")
+		return ctrl.Result{}, nil
+	}
+
+	// Get the cluster api machine
+	machine := &capiv1alpha2.Machine{}
+	machineKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      machineRef.Name,
+	}
+
+	if err := r.Get(ctx, machineKey, machine); err != nil {
+		log.Error(err, "failed to get machine")
+		return ctrl.Result{}, err
+	}
+
+	if machine.Labels[capiv1alpha2.MachineClusterLabelName] == "" {
+		return ctrl.Result{}, errors.New("machine has no associated cluster")
+	}
+
+	// Get the cluster
+	cluster := &capiv1alpha2.Cluster{}
+	clusterKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      machine.Labels[capiv1alpha2.MachineClusterLabelName],
+	}
+	if err := r.Get(ctx, clusterKey, cluster); err != nil {
+		log.Error(err, "failed to get cluster")
```

Comment:
@chuckha general question: would it be sufficient to return the error? Do we need to log it too?

Reply:
yeah, maybe? I don't know. When I was debugging I just logged everything and it worked well, but it might be sufficient to return the error (would be ideal). I'm gonna go read controller-runtime a little more closely and get back to you.

Reply:
controller-runtime will log the error. the main issue is the log source (file:line) is from controller-runtime, not where it actually occurred. That's why Wrap/WithStack is important (assuming the log format prints it out, which it probably won't since it comes from https://github.com/kubernetes-sigs/controller-runtime/blob/ee41a803ccc9c8a308b751433fc35282c716a9ff/pkg/internal/controller/controller.go#L218).

Reply:
We might consider making a logging shim to log errors in a friendly way instead of relying on klogr or zap to do the right thing for us.

(Thank you for finding that, I was just doing the reverse-interface-digging)

Reply:
hmm assuming we can set that log 🤔

Reply:
Moving this to the overall logging theme and tracking here: https://github.com/kubernetes-sigs/cluster-api-provider-docker/issues/153